controllers/sparkcluster_controller.go

```diff
@@ -0,0 +1,529 @@
+)
+
+// LastAppliedConfig is the annotation key used to store object state on owned components.
+const SparkLastAppliedConfig = "distributed-compute-operator.dominodatalab.com/last-applied"
+
+var (
+	// SparkPatchAnnotator applies state annotations to owned components.
+	SparkPatchAnnotator = patch.NewAnnotator(SparkLastAppliedConfig)
+	// SparkPatchMaker calculates changes to state annotations on owned components.
+	SparkPatchMaker = patch.NewPatchMaker(SparkPatchAnnotator)
+)
+
+// SparkClusterReconciler reconciles SparkCluster objects.
+type SparkClusterReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// nolint:dupl
+// SetupWithManager creates and registers this controller with the manager.
+func (r *SparkClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&dcv1alpha1.SparkCluster{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Owns(&corev1.Service{}).
+		Owns(&corev1.ServiceAccount{}).
+		Owns(&appsv1.StatefulSet{}).
+		Owns(&rbacv1.Role{}).
+		Owns(&rbacv1.RoleBinding{}).
+		Owns(&networkingv1.NetworkPolicy{}).
+		Owns(&autoscalingv2beta2.HorizontalPodAutoscaler{}).
+		Complete(r)
+}
+
+const SparkFinalizerName = "distributed-compute.dominodatalab.com/dco-finalizer"
+
+//+kubebuilder:rbac:groups=distributed-compute.dominodatalab.com,resources=sparkclusters,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=distributed-compute.dominodatalab.com,resources=sparkclusters/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=distributed-compute.dominodatalab.com,resources=sparkclusters/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=pods,verbs=list;watch
+//+kubebuilder:rbac:groups="",resources=services;serviceaccounts,verbs=create;update;list;watch
+//+kubebuilder:rbac:groups=apps,resources=deployments,verbs=create;update;list;watch
+//+kubebuilder:rbac:groups=autoscaling,resources=horizontalpodautoscalers,verbs=create;update;delete;list;watch
+//+kubebuilder:rbac:groups=networking.k8s.io,resources=networkpolicies,verbs=create;update;delete;list;watch
+//+kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=roles;rolebindings,verbs=create;update;delete;list;watch
+
+// Reconcile implements state reconciliation logic for SparkCluster objects.
+func (r *SparkClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx, log := r.setLogger(ctx, r.Log.WithValues("sparkcluster", req.NamespacedName))
+
+	log.V(2).Info("reconciliation loop trigged")
+
+	rc := &dcv1alpha1.SparkCluster{}
+	if err := r.Get(ctx, req.NamespacedName, rc); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("resource not found, assuming object was deleted")
+			return ctrl.Result{}, nil
+		}
+
+		log.Error(err, "failed to retrieve resource")
+		return ctrl.Result{}, err
+	}
+
+	err := r.processFinalizers(ctx, rc, log)
```

Comment:
I might exit the reconcile loop with a requeue after this func updates the `SparkCluster` and here's why: Any update to the CR will change the generate version so any other attempts to update it will error out, which will requeue the request as well, but you can avoid the error this way.

Reply:
done, by using the somewhat inelegant (err, bool) return we agreed on

Reply:
definitely somewhat inelegant 😊